README.md
Please refeer to [https://docs.syncity.com/](https://docs.syncity.com/) for the full updated documentation.

## Syncity Toolbox

### Introduction

This [python](https://www.python.org/) toolset allows you to quickly learn how the command line works.

We've included plently of examples to solve specific scenarios while parameters are fully exposed allowing you to fine tune and experiment with them.

You don't necessarly need to develop your solution in python as the command line itself is a simple telnet interface to communicate with. 

### File Structure

- **[syncity.py](https://github.com/Cvedia/syncity-redist/blob/master/syncity.py)**
	- **[cl/](https://github.com/Cvedia/syncity-redist/tree/master/cl)**
		*command line repository*
		- **[head/](https://github.com/Cvedia/syncity-redist/tree/master/cl/head)**
			*all `syncity/scripts` scene setup outputs in command line format*
	- **[syncity/](https://github.com/Cvedia/syncity-redist/tree/master/syncity)**
		- **[scripts/](https://github.com/Cvedia/syncity-redist/tree/master/syncity/scripts)**
			*scripts repository, this is where your custom code should go*
		- **[tools/](https://github.com/Cvedia/syncity-redist/tree/master/syncity/tools)**
			*custom tools assortment for reporting and assessing problems with the simulation*
		- **[helpers.py](https://github.com/Cvedia/syncity-redist/blob/master/syncity/helpers.py)**
			*A collection of helper functions*
		- **[common.py](https://github.com/Cvedia/syncity-redist/blob/master/syncity/common.py)**
			*Telnet and common functions*
		- **[settings_manager.py](https://github.com/Cvedia/syncity-redist/blob/master/syncity/settings_manager.py)**
			*settings singleton controller*
		- **[unity_vars.py](https://github.com/Cvedia/syncity-redist/blob/master/syncity/unity_vars.py)**
			*static unity variables repository*
	- **[etc/](https://github.com/Cvedia/syncity-redist/tree/master/etc)**
			*repository of misc files*
	- **[install/](https://github.com/Cvedia/syncity-redist/tree/master/install)**
		*per platform / implementation easy install scripts*
	- **[tests/](https://github.com/Cvedia/syncity-redist/tree/master/tests)**
		*simulator / sdk pytest scripts*

### How to install

You can clone or download the SDK at [github](https://github.com/Cvedia/syncity-redist), or by command line:

```bash
git clone https://github.com/Cvedia/syncity-redist syncity-redist
```

#### Linux install

Run `install/linux/sdk/install.sh`, this should automatically detect and install all packages you need.


#### Windows install

1. [Download python](https://www.python.org/downloads/windows/) - We recommend python 3+.
2. After install you will have python and pip added to your PATH, then just open a command prompt, go to the folder where you have syncity redistributables and run:

```bash
install/windows/sdk/install.bat
```


### Running

`python syncity.py --help`

Shows you a list of options, switches and so on. What you want to do first is run a simple script, keep in mind that you can run this script from a different machine, as long it has access to the machine running the simulator.

We recommend you running the --help page so you can see a detailed description of all scripts currently available.

#### Warning

By default the telnet based outputs will be placed where the python scripts are running, however, binary outputs, such as images will be placed on the machine where the simulator is running, that's why the SDK has 2 configurable output paths.

### Examples

`python syncity.py --script basic --cooldown 0 -i 192.168.1.75 -p 10200 --record`

This will run script basic (syncity/scripts/basic.py), with no cooldown time, on the machine with ip 192.168.1.75 port 10200 (default) and will record all commands sent to a file.

Depending on the script you run it will also export files, those files will be saved on the machine running the simulator, with the option -o you can configure the path to write those files, this defaults to E:\tmp\

`python syncity.py --script basic --cooldown 0 --keep`

By default, once the script script is completed, it will delete all objects on the scene, effectively reseting it, with --keep all objects will be kept so you can interact with them after the script is completed; This is ideal if you want to experiment with different settings without bothering creating a full scenario.

`python syncity.py -r file.txt`

Instead of running a script, this will just run the commands on file.txt

### Daisy chaining

You can run a combination of scripts, tools and cl in a chain using the SDK, this allows for very complex combinations avoiding you from running several commands at the same time. The order you send the arguments will be respected on the execution, for example:

```bash
python syncity.py \
     -r static_init.cl \
     -s loop1 \
     -t gallery \
     -r static_scenario.cl fixes.cl \
     -s loop2 loop3 \
     -t gallery
```

This will run, in order:
- A local `static_init.cl` file
- A python script `loop1` (syncity/scripts/loop1.py)
- Gallery tool
- Local scripts `static_scenario.cl` followed by `fixes.cl`
- A python script `loop2` (syncity/scripts/loop2.py) followed by `loop3` (syncity/scripts/loop3.py)
- Gallery tool


### Creating your own script / tool

If you want to use our SDK as base for your development you can check `scripts/template.py` and `tools/template.py` as base to create a script or a tool. All you have to do is place the file on it's respective folder and call it from the main controller, like:

`python syncity.py -s my_custom_script`

Where my_custom_script is located at `syncity/scripts/my_custom_script.py`

## Reporting issues

We developed a special tool to generate meaningfull reports by collecting all existing logs, initialization files, hashes, etc. If you happen to have a issue with the simulator running the report tool will probably generate all information required for us to debug the problem.

### Running report tool

`python syncity.py --tool report --binary <path to binary> --assets <path to assets>`

#### Example:

`python syncity.py --tool report --binary E:\syncity\windows-171211 --assets E:\syncity\assets-windows-171211`

#### Notes:

- If you happen to have a large collection of assets you can run this without the `assets` parameter.
- This script will save a `report_XXX.txt` on your `local_path` (defaults to tmp) that you should submit to our support
- No sensitive data will be collected using this tool, keep in mind that all syncity related logs will be sent
- The SDK should work with python 2, but it's support is no longer provided, make sure you're running python 3+.


### Argument list

Here's a list of all currently supported arguments, of course you can find it by running `python syncity.py -h`:

```
SynCity toolbox - v18.04.05.1828
Copyright (c) 2016-2018 CVEDIA PVE Ltd

usage: syncity.py [-h] [-p PORT] [-i IP] [-s SCRIPT [SCRIPT ...]]
                  [-t TOOL [TOOL ...]] [-r RUN [RUN ...]]
                  [--cooldown COOLDOWN] [-o OUTPUT_PATH] [-l LOCAL_PATH] [-q]
                  [-n] [-Z] [-A ASSETS] [--db DB] [--seed_py SEED_PY]
                  [--seed_api SEED_API] [--interactive] [--keep] [-R]
                  [--record [RECORD]] [-b] [-d] [-L LOGLEVEL] [--debug]
                  [--log [LOG]] [--async] [--skip_init] [--skip_disk]
                  [--skip_setup] [-c CONFIG] [--skip_config] [--save_config]
                  [--skip_shutdown] [--test] [--setup_only] [--enable_physics]
                  [--enable_console_log] [--enable_canvas]
                  [--disable_envirosky] [--use_old_depth_buffer] [--flycam]
                  [--nohead] [--carsLimit CARSLIMIT] [--propsLimit PROPSLIMIT]
                  [--signsLimit SIGNSLIMIT] [--treesLimit TREESLIMIT]
                  [--backdropsLimit BACKDROPSLIMIT] [--X_COMP X_COMP]
                  [--Y_COMP Y_COMP] [--Z_COMP Z_COMP] [-v]
                  [--loop_limit LOOP_LIMIT] [--disable_ros]
                  [--lidar_ip LIDAR_IP] [--disable_lidar] [--enable_cube]
                  [--binary BINARY] [--asset_file ASSET_FILE]

This script allows experimentation, recording and playback of scripted events 
using syncity simulator telnet api. We also added some demos that show case 
different specific features of the api.

You can pipe multiple scripts, tools and command line (-r , --run) files on the 
same run, this script will respect the order of the arguments even if they are 
interleaved with each other, this allows for very flexible pipelines, for 
example:

`syncity.py -r path/to/my.cl -s 360 -t gallery -s something something_else -r 
other.cl -s 360 -t gallery`

optional arguments:
  -h, --help            show this help message and exit
  -p PORT, --port PORT  Port to connect, defaults to 10200
  -i IP, --ip IP        IP of syncity simulator
  -s SCRIPT [SCRIPT ...], --script SCRIPT [SCRIPT ...]
                        Defines one or more python scripts to run
  -t TOOL [TOOL ...], --tool TOOL [TOOL ...]
                        Defines one or more tools to run
  -r RUN [RUN ...], --run RUN [RUN ...]
                        Defines one or more command line (.cl) script to run
  --cooldown COOLDOWN   Cooldown after snapshot
  -o OUTPUT_PATH, --output OUTPUT_PATH
                        Defines output path for snapshots, note that this path
                        is relative to the machine running the simulator,
                        defaults to /tmp/
  -l LOCAL_PATH, --local_path LOCAL_PATH
                        Defines local output path for recordings, json
                        exports, etc; This path is relative to the machine
                        running this script, defaults to /tmp/
  -q, --quiet           Quiet mode
  -n, --no_color        Disable color output
  -Z, --abort_on_error  Abort execution on error
  -A ASSETS, --assets ASSETS
                        Defines assets folder name
  --db DB               Defines database folder name, if not set will follow
                        --assets, if --assets is not defined, this value will
                        not be touched
  --seed_py SEED_PY     Defines a seed number for random methods on python
                        layer. Note that this will be treated as a string.
  --seed_api SEED_API   Defines a seed number for random methods on api layer.
  --interactive         Drops to a interactive shell after executing stack or
                        when interrupted.
  --keep                Keep created assets on scene
  -R, --reset           Resets simulator before running anything on stack
  --record [RECORD]     Record commands sent to API using --local_path as
                        output path when no parameters sent, or defines a
                        filename to write to
  -b, --benchmark       Enables benchmarking on API reply time
  -d, --dry_run         Pretends to run, but never send any command thuru
                        telnet
  -L LOGLEVEL, --loglevel LOGLEVEL
                        Minimum log level to display / log.
  --debug               Defines a global debug flag, this will cause a lot of
                        outputs
  --log [LOG]           Log all IOs, if a parameter is specified this will log
                        to a specific file
  --async               Send some telnet commands asyncronously --
                        EXPERIMENTAL
  --skip_init           Disables telnet init sequence
  --skip_disk           Disables disk export completly
  --skip_setup          Skip script setup and go straight to data extraction
  -c CONFIG, --config CONFIG
                        Defines a path/filename for syncity.conf
  --skip_config         Skips SDK default config file
  --save_config         Save sent parameters as SDK config file -- WARNING:
                        This will not save the stack parameters (-r, -s and
                        -t)
  --skip_shutdown       Skips shutdown sequence
  --test                Enables test suite flag
  --setup_only          Runs script setup and exits
  --enable_physics      Enable Physics, mainly affects objects with
                        rigidbodies.
  --enable_console_log  Enable UI console log
  --enable_canvas       Enable client rendering visualization, this is the
                        legacy interface
  --disable_envirosky   Disables Envirosky -- NOT RECOMMENDED
  --use_old_depth_buffer
                        Uses old depth buffer component
  --flycam              Spawns fly cam, controllable via simulator
  --nohead              Disables header outputs on log / record
  --carsLimit CARSLIMIT
                        Spawn cars into scene, defaults to 100
  --propsLimit PROPSLIMIT
                        Spawn props into scene, defaults to 250
  --signsLimit SIGNSLIMIT
                        Spawn signs into scene, defaults to 250
  --treesLimit TREESLIMIT
                        Spawn trees into scene, defaults to 250
  --backdropsLimit BACKDROPSLIMIT
                        Spawn backdrops into scene, defaults to 50
  --X_COMP X_COMP       Spawner X Compensation, defaults to 0
  --Y_COMP Y_COMP       Spawner Y Compensation, defaults to 1
  --Z_COMP Z_COMP       Spawner Z Compensation, defaults to 0
  -v, --version         Shows current SDK version, can be combined with -i and
                        -p to get simulator version


```

## [pytest](https://docs.pytest.org/en/latest/) scripts

This is a test suite for both the SDK and the simulator itself, a typical enduser wouldn't need this.

Keep in mind that some of this tests might take a long time and require specific assets and features your binary might not have.

### Installing

Run the install script under `install/<platform>/test` this will install all required dependencies.

### Running

To run all tests:

```bash
cd tests
python -m pytest -vv
```

You can run a specific test by specifying a test file, for example:

```bash
cd tests
python -m pytest -vv test_export.py
```



CHANGELOG
--------------------------------------------------------------------------------
This change log follows YYMMDD timestamps which are also the version number for
both the SDK and the simulator.
--------------------------------------------------------------------------------

# 180907

- #269804617 RandomProps.Spawners.PropArea now supports a list of assets to be spawned
- #255897254 Custom log path support
- Added endless runner instance id support with configurables
- Misc helper functions changes

# 180906

- SKY segmentation support
- Jumanji: Changes to support sky segmentation
- SDK: Segmentation cameras will now automatically add SKY among with VOID into the lookup table by default. To avoid conflicting with existing colors, SKY is segmented as a orangish color (#FF7B00)

# 180903

- Removed random.choices dependency in favor to random.sample

# 180902

- endless_runner: angularVelocity calculations fixed

# 180901

- endless_runner: angularVelocity configuration fixed

# 180831

- *IMPORTANT*: SDK now requires commentjson module, you can install it by running `python3 -m pip install commentjson`
- endless_runner: Other cars maximum speed can now be configured
- endless_runner: Obstacle cars now have a optional maximum time to live
- endless_runner: Obstacle cars no longer get the main car stuck forever
- endless_runner: Obstacle cars can now be spawned in a managed loop managed
- endless_runner: Misc reconfiguration in obstacle settings
- endless_runner: Reflection probes spawning system been optimized, greatly improving performance; endless_runner with stereo thermal cameras perform 60FPS in 1080TI now.
- endless_runner, jumanji and poc_drone scenes' reflection probes optimized (single probe approach)
- dataset_generation scripts: Replicated optimizations in the endless_runner configurables
- Internal QA tests and devtools are now private git submodules
- SDK's logging and recording functionalities are now enabled by default and stored in logs/ and records/ at SDK folder
- SDK now supports JSON files with comments
- SDK JSON configuration files been all commented, explaining all configurations
- Deprecated --log and --record switches

# 180824

- GPU memory leak issues fixed
- Support to unload assetbundles via API
- Refactoring of poc_drone spawn limits
- Misc changes / fixes in helpers

# 180822

- #265386365: Fixed memory overflow in poc_drone
- #259415293: Fixed FPS in video exports
- Fixed issue with syncity crashing out of memory due excess of cached entities
- Fixed argument passing issue with DataExporter in SDK
- Optimizations and better camera settings for endless_runner scene
- Several minor fixes in SDK scripts

# 180816

- #259735641 File mask support added in the SDK
- Fixed issue with bicycles in jumanji scene
- Added endless runner entropic scene
- Fixed VehicleSpawner naming in endless runner
- Enabled / optimized reflection probes in endless runner and other thermal scenes
- Fix in endless_runner entropy script that would cause bounding boxes to not show in some cases
- New thermal reflection probe shader (fixes problems with thermal cameras + rgb + reflection probes in the same scene)
- New thermal cars, humans and clothing asset bundle packages
- Better thermal camera settings

# 180808

- Fixed critical issue with bounding boxes not being aligned
- Jumanji T-Pose fixes
- DBUpdate tool added
- Fixed issue related to not all parameters are passed thuru the backend
- Added texture randomization example for jumanji
- #259132755 Fixed issue with humans when using the generic spawner
- #259735641 Allow output filename customization
- #260373068 Proper acceleration vectors
- #260125142 Gallery tool now works with new boundingbox types
- #259628624 Clustering of humans / cyclists, etc (endless_runner)
- #260471509 Jumanji cleanup
- #260941684 Fixed missing bounding boxes at random
- #260440779 Fixed issue with alongated bounding boxes in endless runner scene
- Added visibility to bounding box export message

# 180730

- New DataExport module, standarizing all types of data exports into a single component
- Arbitrary data types (float, int, vector3, etc) exports to json support
- Data exports are now timestamped individually, allowing for very precise data synchronization
- Mixed data exports (video, image, data) in the same synchronizer package are now supported
- SDK DataExport function support and deprecation of old methods
- SDK graceful deprecation of VideoExport and ImageExport methods
- SDK misc refactoring
- DataExports can now be paused and resumed at any point
- Forest thermal script changes; Added new data export module support
- Endless runner script changes; Added new data export module support; Added support to export IMU and other vector data;
- Camping script changes
- Fixed Jumanji Scene and added new data export module support
- Fixed issue causing all RandomArea spawners from going into a endless loop when toggling enabled / disabled
- Fixed issues with aerial_cars_depth_thermal_easier and added new data export module support
- Added generic exporters for video and image
- Shopping mall scene reviews
- Fixed issue with missing configurable for streamOutput on DataExport

* Disk module has been deprecated in favor to DataExport
* Segmentation snapshot telnet based json helper methods been deprecated in favor to DataExport, although this still available via telnet
* ROS exports been temporarly disabled in this release as it's being refactored to the ICE bus

# 180723:

- Documentation updated
- Fixed critical issues with endless runner scene
- Fixed critical issues with obsolete components not being properly redirected

# 180720:

- New image export rollout among with json exporter per-frame compatibility
- Segmentation exports now perform 10x faster and can work in asynchronous mode
- Several thermal profiles added
- Added scripts to export images using the new modules
- Partially updated existing scripts and cls to work with the new exporter

syncity.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

from __future__ import division

import sys
import os
import time
import argparse
import textwrap
import platform
import datetime
import syncity
import random

from syncity import common, settings_manager

SYNCITY_VERSION = '18.09.07.0331'
SIMULATOR_MIN_VERSION = '18.04.23.0000'

print ('SynCity toolbox - v{}\nCopyright (c) 2016-{} CVEDIA PVE Ltd\n'.format(SYNCITY_VERSION, datetime.date.today().year))

if sys.version_info[0] < 3:
	print ('*** WARNING: You\'re using a old version of python that is not maintained by this SDK, some functions might fail. Run at your own risk! Please use python 3.6+\n\n')

parser = argparse.ArgumentParser(
	formatter_class=argparse.RawDescriptionHelpFormatter,
	description=textwrap.dedent('''\
This script allows experimentation, recording and playback of scripted events using syncity simulator telnet api. We also added some demos that show case different specific features of the api.

You can pipe multiple scripts, tools and command line (-r , --run) files on the same run, this script will respect the order of the arguments even if they are interleaved with each other, this allows for very flexible pipelines, for example:

`syncity.py -r path/to/my.cl -s 360 -t gallery -s something something_else -r other.cl -s 360 -t gallery`
'''),
	epilog=textwrap.dedent('Scripts available:\n\n{}\n\nTools available:\n\n{}\n\nFull documentation available in: https://docs.syncity.com'.format(common.modulesHelp('scripts'), common.modulesHelp('tools')))
)

parser.add_argument('-p', '--port', type=int, default=10200, help='Port to connect, defaults to 10200')
parser.add_argument('-i', '--ip', default='127.0.0.1', help='IP of syncity simulator')

# pipeable arguments
parser.add_argument('-s', '--script', default=None, action='append', nargs='+', help='Defines one or more python scripts to run')
parser.add_argument('-t', '--tool', default=None, action='append', nargs='+', help='Defines one or more tools to run')
parser.add_argument('-r', '--run', type=argparse.FileType('r'), action='append', nargs='+', help='Defines one or more command line (.cl) script to run')

parser.add_argument('--cooldown', type=int, default=0, help='Cooldown after snapshot')

# platform specific
if platform.system() == 'Windows':
	parser.add_argument('-o', '--output', default='E:\\tmp\\', help='Defines output path for snapshots, note that this path is relative to the machine running the simulator, defaults to E:\\tmp\\', dest='output_path')
	parser.add_argument('-l', '--local_output', default='E:\\tmp\\', action=common.readableDir, help='Defines local output path for recordings, json exports, etc; This path is relative to the machine running this script, defaults to E:\\tmp\\', dest='local_path')
else:
	parser.add_argument('-o', '--output', default='/tmp/', help='Defines output path for snapshots, note that this path is relative to the machine running the simulator, defaults to /tmp/', dest='output_path')
	parser.add_argument('-l', '--local_output', default='/tmp/', action=common.readableDir, help='Defines local output path for recordings, json exports, etc; This path is relative to the machine running this script, defaults to /tmp/', dest='local_path')

parser.add_argument('--log_output', default=None, action=common.readableDir, help='Defines a path to write recordings and log files. This overrides the defaults logs/ and records/', dest='log_path')

# misc
parser.add_argument('-q', '--quiet', action='store_true', default=False, help='Quiet mode')
parser.add_argument('-n', '--no_color', action='store_true', default=False, help='Disable color output')
parser.add_argument('-Z', '--abort_on_error', action='store_true', default=False, help='Abort execution on error')
parser.add_argument('-A', '--assets', help='Defines assets folder name')
parser.add_argument('--db', help='Defines database folder name, if not set will follow --assets, if --assets is not defined, this value will not be touched')
parser.add_argument('--seed_py', help='Defines a seed number for random methods on python layer. Note that this will be treated as a string.', default=None)
parser.add_argument('--seed_api', type=int, help='Defines a seed number for random methods on api layer.', default=None)
parser.add_argument('--interactive', action='store_true', default=False, help='Drops to a interactive shell after executing stack or when interrupted.')
parser.add_argument('--keep', action='store_true', default=False, help='Keep created assets on scene')
parser.add_argument('-R', '--reset', action='store_true', default=False, help='Resets simulator before running anything on stack')
parser.add_argument('-b', '--benchmark', action='store_true', default=False, help='Enables benchmarking on API reply time')
parser.add_argument('-d', '--dry_run', action='store_true', default=False, help='Pretends to run, but never send any command thuru telnet')
parser.add_argument('-L', '--loglevel', default='DEBUG', help='Minimum log level to display / log.')
parser.add_argument('--debug', action='store_true', default=False, help='Defines a global debug flag, this will cause a lot of outputs')
parser.add_argument('--async', action='store_true', default=False, help='Send some telnet commands asyncronously -- EXPERIMENTAL')

parser.add_argument('--skip_init', action='store_true', default=False, help='Disables telnet init sequence')
parser.add_argument('--skip_export', action='store_true', default=False, help='Disables data exports completly')
parser.add_argument('--skip_setup', action='store_true', default=False, help='Skip script setup and go straight to data extraction')
parser.add_argument('--skip_queue', action='store_true', default=False, help='Skips all queue waits, this might cause asyncronous states for exports -- NOT RECOMMENDED')

parser.add_argument('-c', '--config', default=None, help='Defines a config file to be loaded, overwriting / appending to arbitrary settings values')
# parser.add_argument('--skip_config', action='store_true', default=False, help='Skips SDK default config file')
# parser.add_argument('--save_config', action='store_true', default=False, help='Save sent parameters as SDK config file -- WARNING: This will not save the stack parameters (-r, -s and -t)')
parser.add_argument('-O', '--options', default=None, action='append', nargs='+', help='Defines a path to a json file to overload / overwrite settings.options to be used as parameters on scripts globally. You can concatenate files, they will be merged into a single option stack. Note that the order you define the parameters of this function will affect how they are merged.')
parser.add_argument('--layout', default=None, help='Path to layout file, this file must be accessible to the simulator')

parser.add_argument('--skip_shutdown', action='store_true', default=False, help='Skips shutdown sequence')
parser.add_argument('--test', action='store_true', default=False, help='Enables test suite flag')

parser.add_argument('--setup_only', action='store_true', default=False, help='Runs script setup and exits')
parser.add_argument('--enable_physics', action='store_true', default=False, help='Enable Physics, mainly affects objects with rigidbodies.')
# parser.add_argument('--enable_console_log', action='store_true', default=False, help='Enable UI console log')
parser.add_argument('--enable_canvas', action='store_true', default=False, help='Enable client rendering visualization, this is the legacy interface')
parser.add_argument('--disable_envirosky', action='store_true', default=False, help='Disables Envirosky -- NOT RECOMMENDED')
parser.add_argument('--disable_reflections', action='store_true', default=False, help='Disables Reflection probes')
parser.add_argument('--disable_logging', action='store_true', default=False, help='Disables logging')
parser.add_argument('--disable_records', action='store_true', default=False, help='Disables session recording')
parser.add_argument('--fake_version', default=False, help='Overrides the current SDK version')

parser.add_argument('--cache_limit', type=int, default=0, help='Defines how large the asset cache pool should be. The higher you set this more GPU memory will be used.')

parser.add_argument('--use_old_depth_buffer', action='store_true', default=False, help='Uses old depth buffer component')
parser.add_argument('--flycam', action='store_true', default=False, help='Spawns fly cam, controllable via simulator')
parser.add_argument('--nohead', action='store_true', default=False, help='Disables header outputs on log / record')

parser.add_argument('-T', '--queue_threshold', type=int, default=0, help='Defines a threshold to wait for simulator queue, if queue is higher, SDK will wait until it\'s lower than the threshold.')

# TODO: To be deprecated
parser.add_argument('--carsLimit', type=int, default=50, help='Spawn cars into scene, defaults to 100')
parser.add_argument('--propsLimit', type=int, default=250, help='Spawn props into scene, defaults to 250')
parser.add_argument('--signsLimit', type=int, default=250, help='Spawn signs into scene, defaults to 250')
parser.add_argument('--treesLimit', type=int, default=250, help='Spawn trees into scene, defaults to 250')
parser.add_argument('--backdropsLimit', type=int, default=50, help='Spawn backdrops into scene, defaults to 50')

parser.add_argument('--X_COMP', type=float, default=0, help='Spawner X Compensation, defaults to 0')
parser.add_argument('--Y_COMP', type=float, default=1, help='Spawner Y Compensation, defaults to 1')
parser.add_argument('--Z_COMP', type=float, default=0, help='Spawner Z Compensation, defaults to 0')

parser.add_argument('-v', '--version', action='store_true', default=False, help='Shows current SDK version, by explicitly combining with -i, --ip and / or -p, --port will also return the simulator version')

# DEPRECATED - Those options don't do anything, they're here just as backwards compatibility
parser.add_argument('--skip_disk', action='store_true', default=False, help='DEPRECATED, Disables disk export completly')
parser.add_argument('--log', default=False, nargs='?', help='DEPRECATED, logs are always enabled.')
parser.add_argument('--record', default=False, nargs='?', help='DEPRECATED, recordings are always enabled.')

scripts_parser = parser.add_argument_group('Scripts specific options')
common.modulesArgs('scripts', parser=scripts_parser)

tools_parser = parser.add_argument_group('Tool specific options')
common.modulesArgs('tools', parser=tools_parser)

args = parser.parse_args()

# WARNING: settings.<value> set here it might be overwritten when we iterate args.__dict__ if you have conflicts
settings = syncity.settings_manager.Singleton()
settings._start = time.time()
settings._tn = None

settings._version = settings.fake_version if settings.fake_version != False else SYNCITY_VERSION
settings._simulator_min_version = SIMULATOR_MIN_VERSION

settings._root = os.path.dirname(os.path.realpath(__file__))
settings._shutdown = False
settings._interactive = False
settings.log = False if args.__dict__['disable_logging'] == True else None
settings.record = False if args.__dict__['disable_records'] == True else None

common.init()

settings._stack = common.findArgOrder([
	{'id': 'run', 'args': ['-r', '--run']},
	{'id': 'script', 'args': ['-s', '--script']},
	{'id': 'tool', 'args': ['-t', '--tool']}
])

settings._stack_size = len(settings._stack)

for k in args.__dict__:
	if k in ['log', 'record']: # skip DEPRECATED options
		continue
	# print('{}: {}'.format(k, args.__dict__[k]))
	settings[k] = args.__dict__[k]

settings._root = os.path.dirname(os.path.realpath(sys.argv[0]))

if settings.config != None:
	common.loadConfig()
if settings.options != None:
	common.loadOptions()
if platform.system() == 'Windows':
	if settings.local_path.find('/') != -1:
		if settings.local_path[-1:] != '/':
			settings.local_path = settings.local_path + '/'
	elif settings.local_path[-1:] != '\\':
		settings.local_path = settings.local_path + '\\'
else:
	if settings.local_path[-1:] != '/':
		settings.local_path = settings.local_path + '/'

if settings._stack_size == 0 and settings.version == False:
	common.output('Nothing to run, aborting!', 'ERROR')
	sys.exit(0)
elif settings.version == True:
	if common.argExists(['-i', '--ip', '-p', '--port']) == True:
		common.initTelnet(settings.ip, settings.port)
	else:
		common.output('SDK Version: {}'.format(settings._version))
	sys.exit(0)

common.init2()

if settings.seed_py != None:
	random.seed(settings.seed_py)

# setup telnet connection
if settings.run != None or settings.script != None:
	if getattr(settings, "async"):
		common.output('Telnet mode set to ASYNCRONOUS')
		settings.force_sync = False
	else:
		common.output('Telnet mode set to SYNCRONOUS')
		settings.force_sync = True
	
	common.initTelnet(settings.ip, settings.port)
	
	if settings.reset:
		common.resetSimulator()
	
	common.flushBuffer()

settings._idx = { 'run': 0, 'script': 0, 'tool': 0 }
ran = 1

for s in settings._stack:
	common.output('[{}% {}/{}] Running stack: {}'.format(round(100 * (ran / settings._stack_size), 2), ran, settings._stack_size, s))
	
	if s == 'run':
		for subject in settings.run[settings._idx['run']]:
			m = 'Running command line script {} {}...'.format(subject.name, common.md5(subject.name))
			common.output(m)
			common.recordS('// {}'.format(m))
			
			common.runCL(subject.name)
		
		settings._idx['run'] += 1
	elif s == 'script':
		for subject in settings.script[settings._idx['script']]:
			if subject[-3:] == '.py':
				subject = subject[:-3]
			m = 'Running python script: {} {}'.format(subject, common.md5('syncity/scripts/{}.py'.format(subject)))
			common.output(m)
			common.recordS('// {}'.format(m))
			
			# track objects created by script to remove them from scene later on
			settings._obj = []
			settings._seqSave_i = 1
			
			# this should work with both python 2.7 and 3+
			import_script = __import__('syncity.scripts.{}'.format(subject), fromlist=['syncity.scripts'])
			
			if not settings.dry_run:
				try:
					mv = import_script.minVersion()
					if common.versionCompare(settings._simulator_version, mv, '<'):
						common.output('This script requires simulator newer than {}, you\'re running {}. Aborting!'.format(mv, settings._simulator_version), 'OBSOLETE')
						sys.exit(1)
				except:
					pass
				
				try:
					mv = import_script.maxVersion()
					if common.versionCompare(settings._simulator_version, mv, '>'):
						common.output('This script requires simulator older than {}, you\'re running {}. Aborting!'.format(mv, settings._simulator_version), 'OBSOLETE')
						sys.exit(1)
				except:
					pass
			
			import_script.run()
			common.flushBuffer()
		
		settings._idx['script'] += 1
	elif s == 'tool':
		for subject in settings.tool[settings._idx['tool']]:
			if subject[-3:] == '.py':
				subject = subject[:-3]
			m = 'Running tool: {} {}'.format(subject, common.md5('syncity/tools/{}.py'.format(subject)))
			common.output(m)
			common.recordS(m)
			
			import_tool = __import__('syncity.tools.{}'.format(subject), fromlist=['syncity.tools'])
			import_tool.run()
		
		settings._idx['tool'] += 1
	else:
		raise 'Trying to run unknown stack type: {}'.format(s)
	
	ran += 1

common.output('Stack completed.')
